conventions/c++.rst

```diff
@@ -0,0 +1,414 @@
+  tends to lead one to write ``SomeClass foo();`` which is a function
+  prototype, and not equivalent to the variable definition ``SomeClass
+  foo``.  However, note that ``new SomePODType()`` is *not* the same
+  as ``new SomePODType`` (if SomePODType is a Plain Old Data type) -
+  the former will zero-initialise scalar members of SomePODType.
+
+* RTTI (``dynamic_cast<>``, ``typeid``, ``std::typeinfo``): Needing to
+  use RTTI features in the library most likely indicates a design
+  flaw, and you should avoid use of these features.  Where necessary,
+  you can use a technique similar to
+  ``Database::as_networkdatabase()`` to replace ``dynamic_cast<>``.
+
+* ``using namespace std;`` and ``using std::XXX;`` - it's OK to use
+  these in applications, library code, and internal library headers.
+  But in externally visible headers (such as anything included by
+  ``#include <xapian.h>``) you *MUST* use explicit ``std::``
+  qualifiers - it's not acceptable to pull anything from namespace std
+  into the namespace of an application which uses Xapian.
+
+* Use C++ style casts (``static_cast<>``, ``reinterpret_cast<>``, and
+  ``const_cast<>``) or constructor-syntax (e.g. ``double(value)``) in
+  preference to C style casts.  The syntax of the C++ casts is ugly,
+  but they do make the intent much clearer which is definitely a good
+  thing, and they avoid issues such as casting away const when you
+  only meant to cast the type of a pointer.
+
+* ``std::pair<>`` with an STL class as one (or both) of the members
+  can produce very long symbols (over 4KB!) after name mangling - long
+  enough to overflow the size limits of some vendor compilers or
+  toolchains (so this can affect GCC if it is using the system ld or
+  as).  Even where the compiler works, the symbol bloat in an
+  unstripped build is probably best avoided, so it's preferable to use
+  a simple two member struct instead.  The code is probably more
+  readable anyway, and easier to extend if more members are needed
+  later.
+
+* We try to avoid putting the full definition of virtual methods in header
+  files.  This is because current compilers can't (as far as we know) inline
+  virtual methods, so putting the definition in the header file simply slows
+  down compilation (and, because method definitions often require further
+  header files to be included, this can result in many more files needing
+  recompilation after a change to a header file than is really necessary).
+  Just put the declaration in the header file, and put the definition in a .cc
+  file with the same basename.
+
+
+Efficient use of ``std::string``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* When passing an empty string to a method expecting ``const
+  std::string &`` prefer ``std::string()`` to ``""`` or
+  ``std::string("")`` as the first form is more likely to directly use
+  a special "empty string representation" (it does with GCC at least).
```

Comment:
The last part is no longer true since C++11 outlawed ref-counted std::string implementations.

I think last time I checked, `std::string("")` (explicit or implicit) resulted in the same code as `std::string()` for GCC, but was less efficient with clang, so `std::string()` for an empty `std::string` is still reasonable advice.